src-input/duk_bi_encoding.c

```diff
@@ -0,0 +1,333 @@
+	if (enc_ctx.lead != 0x0000) {
+		/* unpaired high surrogate at end of string */
+		enc_ctx.out += duk_unicode_encode_xutf8(DUK_UNICODE_CP_REPLACEMENT_CHARACTER, enc_ctx.out);
+	}
+#if defined(DUK_USE_BUFFEROBJECT_SUPPORT)
+	duk_push_buffer_object(ctx, -1, 0, (duk_size_t) (enc_ctx.out - output), DUK_BUFOBJ_UINT8ARRAY);
+#else
+	/* shrink buffer to include only the encoded text */
+	duk_resize_buffer(ctx, -1, (duk_size_t) (enc_ctx.out - output));
+#endif
+	return 1;
+}
+
+DUK_INTERNAL duk_ret_t duk_bi_textdecoder_constructor(duk_context *ctx) {
+	duk__decode_context *dec_ctx;
+	duk_bool_t fatal = 0;
+	duk_bool_t ignore_bom = 0;
+	const char *label = "utf-8";
+	duk_idx_t nargs;
+
+	nargs = duk_get_top(ctx);
+	if (nargs >= 1) {
+		label = duk_to_string(ctx, 0);
+		/* FIXME: do something with label */
+	}
+	if (nargs >= 2 && !duk_is_null_or_undefined(ctx, 1)) {
+		if (duk_get_prop_string(ctx, 1, "fatal")) {
+			fatal = duk_to_boolean(ctx, -1);
+		}
+		if (duk_get_prop_string(ctx, 1, "ignoreBOM")) {
+			ignore_bom = duk_to_boolean(ctx, -1);
+		}
+	}
+
+	dec_ctx = (duk__decode_context *) duk_alloc(ctx, sizeof(duk__decode_context));
+	dec_ctx->fatal = fatal;
+	dec_ctx->ignore_bom = ignore_bom;
+	duk__utf8_decode_init(dec_ctx);
+
+	duk_push_this(ctx);
+	duk_push_pointer(ctx, (void *) dec_ctx);
+	duk_put_prop_string(ctx, -2, "\xff" "Context");
+	return 0;
+}
+
+DUK_INTERNAL duk_ret_t duk_bi_textdecoder_prototype_encoding_getter(duk_context *ctx) {
+	duk_push_string(ctx, "utf-8");
+	return 1;
+}
+
+DUK_INTERNAL duk_ret_t duk_bi_textdecoder_prototype_fatal_getter(duk_context *ctx) {
+	duk__decode_context *dec_ctx;
+
+	duk_push_this(ctx);
+	duk_get_prop_string(ctx, -1, "\xff" "Context");
+	dec_ctx = (duk__decode_context *) duk_require_pointer(ctx, -1);
+	duk_push_boolean(ctx, dec_ctx->fatal);
+	return 1;
+}
+
+DUK_INTERNAL duk_ret_t duk_bi_textdecoder_prototype_ignorebom_getter(duk_context *ctx) {
+	duk__decode_context *dec_ctx;
+
+	duk_push_this(ctx);
+	duk_get_prop_string(ctx, -1, "\xff" "Context");
+	dec_ctx = (duk__decode_context *) duk_require_pointer(ctx, -1);
+	duk_push_boolean(ctx, dec_ctx->ignore_bom);
+	return 1;
+}
+
+DUK_INTERNAL duk_ret_t duk_bi_textdecoder_prototype_decode(duk_context *ctx) {
+	duk_uint8_t *data = (duk_uint8_t *) "";
+	duk__decode_context *dec_ctx;
+	duk_size_t length = 0;
+	duk_idx_t nargs;
+	duk_bool_t stream = 0;
+	duk_codepoint_t codepoint;
+	int ret;
+	duk_uint8_t *output;
+	duk_uint8_t *in, *out;
+
+	nargs = duk_get_top(ctx);
+	if (nargs >= 1) {
+		data = (duk_uint8_t *) duk_require_buffer_data(ctx, 0, &length);
+	}
+	if (nargs >= 2 && !duk_is_null_or_undefined(ctx, 1)) {
+		if (duk_get_prop_string(ctx, 1, "stream")) {
+			stream = duk_to_boolean(ctx, -1);
+		}
+	}
+	duk_push_this(ctx);
+	duk_get_prop_string(ctx, -1, "\xff" "Context");
+	dec_ctx = (duk__decode_context *) duk_require_pointer(ctx, -1);
+
+	/* FIXME: 100% USDA Choice Grade A memory pork */
+	output = (duk_uint8_t *) duk_push_fixed_buffer(ctx, length * 3);
```

Comment:
Had some time to reflect on this and it should be good enough for now to keep the allocation as `3 * len` in the general case, but set a lower bound on the allocation of 6 bytes.  In the worst case you get 6 bytes of output for 1 byte of input, but that's a corner case--in general you can never get more than three times the number of bytes you put in because codepoints are emitted as soon as possible.  The only time it will expand by 6x is when the caller sends the final byte of a 4-byte sequence--and *only* that byte.

We can optimize for space later, but this simple fix should be good enough for master for now. :)